### ISSUE TITLE
Surround does not handle alternate input methods

### ISSUE THREAD

**Issue by eproxus** [NeoVintageous/NeoVintageous#842]
**Describe the bug**  
When using surround to change single quotes to backticks with an alternate keyboard layout that uses Alt the text `<M-'>` is inserted instead.

**To Reproduce**  
1. Add some text: `'foo'`
2. Type <code>cs'`</code>

**Expected behavior**  
The text should be <code>`foo`</code>

**Actual behavior**  
The text is `<M-'>foo<M-'>`

**Versions**  
`startup, version: 4136 osx x64 channel: dev`

**Additional Information**  

A backtick is inserted on my keyboard layout with <kbd>AltGr</kbd>+<kbd>'</kbd>.

**Comment by gerardroche** [comment]
Can you give me some details on what input method you're using and language. I wasn't able to reproduce this issue.

**Comment by eproxus** [comment]
I'm using a custom Swedish Dvorak keyboard layout (but I don't suspect that is the cause, there are probably other keyboard layouts exhibiting the same problem).

Here is my layout: [svdvorak.bundle.zip](https://github.com/NeoVintageous/NeoVintageous/files/9662520/svdvorak.bundle.zip). Place it in `/Library/Keyboard Layouts/` (as root) or in `~/Library/Keyboard Layouts/`. You can use the keyboard viewer to see which key is the back tick (_System Settings_ → _Keyboard_ → _Input Sources_ → _Show Input menu in menu bar_).

If you don't want to install a random layout file from the Internet, I can reproduce it with the macOS built-in Swedish layout by using the <kbd>AltGr</kbd>+<kbd>Q</kbd> combination (which should yield a `•` character, but gets inserted as `<M-q>foo<M-q>`).

**Comment by gerardroche** [comment]
Yeah, I don't think it's the keyboard layout. I just need a way to reproduce it so I can figure out what needs to be done to resolve it, if possible.

**Comment by eproxus** [comment]
Let me know if I can be of any more help.

**Comment by gerardroche** [comment]
I can reproduce the issue using most layouts.

For example, in English (UK)

1. `cs'` 
2. `alt+{char}`

When a regular character is pressed Sublime emits a character event:

```
chr evt: ' (0x27)
```

In this case nv receives the character `'`.

Some keys presses emit both a key *and* character event e.g. pressing `altgr+q`:

```
key evt: altgr+q
chr evt: @ (0x40)
```

In this case nv receives the `@` character.

But some key presses only emit a key event e.g. pressing most `alt+{char}` combinations:

```
key evt: alt+a
```

In this case nv receives `<M-a>`. There is no way to know what character the key represents in the selected input language.

I guess this is the same issue as [Disable Input Method when exiting Insert Mode](https://github.com/NeoVintageous/NeoVintageous/issues/374) and would be solved by porting a input switcher like [vim-xkbswitch](https://github.com/lyokha/vim-xkbswitch).

**Comment by eproxus** [comment]
Not 100% sure if this is related, but macro recording is also broken like this. E.g. recording a macro with `qadawq` and then trying to use `@a` to replay it doesn't work.

If it's a separate issue, let me know and I'll open a new one.

**Comment by eproxus** [comment]
While trying this again in the hope that it would have been fixed in 1.28.0 (because of some macro changes) I noticed the following:

1. In insert mode, Sublime logs _both_ the key and characters event:
   ```
   key evt: alt+2
   chr evt: @ (0x40)
   ```
2. Once back in normal mode, it just logs the key event:
   ```
   key evt: alt+2
   ```

When starting Sublime in safe mode, scenario 1 is always active. Is there anything NeoVintageous has done in order to implement normal mode that throws away the char event perhaps? Could it be added back in terms of interpreting keys somehow? (I'm thinking of a way to solve this that doesn't involve implementing all of vim-xkbswitch)

**Comment by gerardroche** [comment]
Neo just hooks into the key by defining a keymap which feeds neo the key e.g.:

```
    {"keys": ["g"],         "command": "nv_feed_key", "args": {"key": "g"},        "context": [{"key": "vi_command_mode_aware"}]},
```

When **g** is press neo receives **g**.

```
    {"keys": ["enter"],     "command": "nv_feed_key", "args": {"key": "<cr>"},         "context": [{"key": "vi_command_mode_aware"}, {"key": "nv_handle_key", "operand": "<CR>"}]},
```

Here neo receives `<cr>` when you press enter.

This means you can create kemaps for characters not supported by neo. For example, neo doesn't have i18n support, so it doesn't support `д`. If that character is supposed to be equivalent to `d`, a vim mapping won't work well because this is a larger i18n issue. Instead you could "trick" neo with this keymap:

```
    {"keys": ["д"],         "command": "nv_feed_key", "args": {"key": "d"},        "context": [{"key": "vi_command_mode_aware"}]},
```

Now when you press `д` neo will receive `d` and operate as normally as though `d` has been pressed.

Now, mappings might not work as expected because consider

```
nnoremap д x
```

Neo doesn't actually what `д` is so the mapping would be useless. You would have to still use `d` in the mapping. 


```
nnoremap d x
```

Now when you press `d` or `д` neo will receive `d` and then mapping is resolved which then means `x`.

This was noted in https://github.com/NeoVintageous/NeoVintageous/issues/821#issuecomment-1558690407.


I'm still trying to figure out how best to improve i18n support. My current thinking is to provide neovintageous keymap packages for different languages. Each package would have keymaps file which maps the unknown characters to the known characters in neo. And a map of those same characters for the rc file mappins so that you could in fact do `nnoremap д x`.

For example, the keymap file:

```
{"keys": ["д"],         "command": "nv_feed_key", "args": {"key": "d"},        "context": [{"key": "vi_command_mode_aware"}]},
// ...
```

The rc file map:
```
{
    "д": "d",
   //...
}
```

Each language package would have something like that.

Keyboard layouts are little different. You use Dvorak, so check the experimental NeoVintageous Dvorak mapping pack https://github.com/gerardroche/NeoVIntageousDvorak 

Currently you need to copy those mappings into your own mapping. Nothing is auto loaded yet. But I'm looking into a way to let uses load existing mappings files like the Dvorak and [Colemak](https://github.com/gerardroche/NeoVIntageousColemak) **before** the user rc file is loaded.

**Comment by eproxus** [comment]
@gerardroche Thanks for the explanation!

For a workaround for now, I added this my Sublime keymap which makes NeoVintageous behave as expected:
```json
    {
        "keys": ["alt+'"],
        "command": "nv_feed_key",
        "args": {"key": "`"},
        "context": [{"key": "vi_command_mode_aware"}]
    },
    {
        "keys": ["alt+2"],
        "command": "nv_feed_key",
        "args": {"key": "@"},
        "context": [{"key": "vi_command_mode_aware"}]
    },
```

If some other keys is broken, use `sublime.log_input(True)` in the Sublime Console and figure out what Sublime sees and add a mapping for that to whatever you want NeoVintageous to see.
